src/Config.js

```diff
@@ -0,0 +1,212 @@
+// @flow
+import pkgUp from 'pkg-up';
+import detectIndent from 'detect-indent';
+import parseJson from 'parse-json';
+import type { JSONValue, DependencySet } from './types';
+import * as fs from './utils/fs';
+import * as path from 'path';
+import multimatch from 'multimatch';
+import { PError } from './utils/errors';
+
+async function getPackageStack(cwd: string) {
+  let stack = [];
+  let searching = cwd;
+
+  while (searching) {
+    let filePath = await Config.findConfigFile(searching);
+
+    if (filePath) {
+      let config = await Config.init(filePath);
+      stack.unshift({ filePath, config });
+      searching = path.dirname(path.dirname(filePath));
+    } else {
+      searching = null;
+    }
+  }
+
+  return stack;
+}
+
+function toArrayOfStrings(value: JSONValue, message: string) {
+  if (!Array.isArray(value)) {
+    throw new PError(message);
+  }
+
+  return value.map(item => {
+    if (typeof item !== 'string') {
+      throw new PError(message);
+    } else {
+      return item;
+    }
+  });
+}
+
+function toObjectOfStrings(value: JSONValue, message: string) {
+  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
+    throw new PError(message);
+  }
+
+  let safeRef = value;
```

Comment:
Is this actually required? I'm inclined to assume yes, but just wondering why? If we replaced safeRef with value everywhere here, I would think this would be just as safe?

Reply:
It avoids having a refinement invalidation. Flow assumes all functions could mutate values behind the scenes. Re-assigning the value means that I'm not changing the binding to be another value.